data/sql_setup/prepare_schedule_a.sql

```diff
@@ -14,49 +14,86 @@ create index on ofec_sched_a_queue_old (timestamp);
-            insert into ofec_sched_a_queue_new values (new.*, timestamp, two_year_transaction_period_new);
+            delete from ofec_sched_a_queue_new where sub_id = view_row.sub_id;
+            insert into ofec_sched_a_queue_new values (view_row.*, timestamp, two_year_transaction_period_new);
         end if;
 
         return new;
     elsif tg_op = 'UPDATE' then
-        two_year_transaction_period_new = get_transaction_year(new.contb_receipt_dt, new.rpt_yr);
-        two_year_transaction_period_old = get_transaction_year(old.contb_receipt_dt, old.rpt_yr);
+        select into view_row * from fec_vsum_sched_a_vw where sub_id = new.sub_id;
+        two_year_transaction_period_new = get_transaction_year(new.contb_receipt_dt, view_row.rpt_yr);
 
         if two_year_transaction_period_new >= start_year then
-            delete from ofec_sched_a_queue_new where sub_id = new.sub_id;
-            delete from ofec_sched_a_queue_old where sub_id = old.sub_id;
-            insert into ofec_sched_a_queue_new values (new.*, timestamp, two_year_transaction_period_new);
-            insert into ofec_sched_a_queue_old values (old.*, timestamp, two_year_transaction_period_old);
+            delete from ofec_sched_a_queue_new where sub_id = view_row.sub_id;
+            insert into ofec_sched_a_queue_new values (view_row.*, timestamp, two_year_transaction_period_new);
         end if;
 
         return new;
-    elsif tg_op = 'DELETE' then
-        two_year_transaction_period_old = get_transaction_year(old.contb_receipt_dt, old.rpt_yr);
+    end if;
+end
+$$ language plpgsql;
+
+-- Create trigger to maintain Schedule A queues deletes and updates
+-- These happen before a row is removed/updated so that we can leverage pulling
+-- the new record information from the view itself, which contains the data in
+-- the structure that our tables expect it to be in.
+create or replace function ofec_sched_a_delete_update_queues() returns trigger as $$
+declare
+    start_year int = TG_ARGV[0]::int;
+    timestamp timestamp = current_timestamp;
+    two_year_transaction_period_new smallint;
+    two_year_transaction_period_old smallint;
+    view_row fec_vsum_sched_a_vw%ROWTYPE;
+begin
+    if tg_op = 'DELETE' then
+        select into view_row * from fec_vsum_sched_a_vw where sub_id = old.sub_id;
+        two_year_transaction_period_old = get_transaction_year(view_row.contb_receipt_dt, view_row.rpt_yr);
 
         if two_year_transaction_period_old >= start_year then
-            delete from ofec_sched_a_queue_old where sub_id = old.sub_id;
-            insert into ofec_sched_a_queue_old values (old.*, timestamp, two_year_transaction_period_old);
+            delete from ofec_sched_a_queue_old where sub_id = view_row.sub_id;
+            insert into ofec_sched_a_queue_old values (view_row.*, timestamp, two_year_transaction_period_old);
         end if;
 
         return old;
+    elsif tg_op = 'UPDATE' then
```

Comment:
The `update` operation is defined in both trigger functions because one is associated with a `BEFORE` trigger and the other is associated with an `AFTER` trigger.  The reason behind this is because we're leveraging the state of the data as it appears in the view itself to make sure we're pulling all information.

In the case of an `update` operation, we want the state of the view before the change to the record takes place and after the fact in order to accurately capture the record information so that it is processed correctly (namely the `two_year_transaction_period` field, among other things).  The check for the `update` operation in the `BEFORE` trigger captures the state of the record as it exists prior to any change and inserts it into the old queue for deletion.  The check in the `AFTER` trigger captures the state of the record after the change has taken place and inserts it into the new queue for insertion.

I tried to simplify this as much as possible but the reliance on the view for the best representative state of information for all of this data made this a bit difficult.  That said, this allows us to not have to worry as much about what happens with the source data as long as these views continue to exist and provide us with the information we need in the structure we expect.

Please let me know if anything is still unclear!